### PR TITLE
COMMON: Switch from getDataFork to openFileOrDataFork

### PR DESCRIPTION
**Pull request by phcoder**

### CODE REVIEW

File: base/commandLine.cpp [hunk 1]
```diff
@@ -1444,7 +1444,8 @@ static void calcMD5Mac(Common::Path &filePath, int32 length) {
 	if (!macResMan.open(fileName, dir)) {
 		printf("Mac resource file '%s' not found or could not be open\n", filePath.toString(nativeSeparator).c_str());
 	} else {
-		if (!macResMan.hasResFork() && !macResMan.hasDataFork()) {
+		Common::ScopedPtr<Common::SeekableReadStream> dataFork(Common::MacResManager::openFileOrDataFork(fileName, dir));
+		if (!macResMan.hasResFork() && !dataFork) {
```
Comment (sev-) [inline base/commandLine.cpp:1448]: Isn't this mean that if a file will be in plain format, `dataFork` will always be not null here?

File: engines/advancedDetector.cpp [hunk 2]
```diff
@@ -566,7 +566,7 @@ static bool getFilePropertiesIntern(uint md5Bytes, const AdvancedMetaEngine::Fil
 		if (fileProps.size != 0)
 			return true;
 
-		Common::SeekableReadStream *dataFork = macResMan.getDataFork();
+		Common::SeekableReadStream *dataFork = Common::MacResManager::openFileOrDataFork(fname, fileMapArchive);
```
Comment (sev-) [inline engines/advancedDetector.cpp:569]: Same question. Does it mean that when a file is plain and could be missing the required resource fork, we will happily report it as data fork?

E.g. in this particular case we need to check explicitly that the data _fork_ is present

**Comment by phcoder** [comment]
> I am afraid that we are oversimplifying things.
> 
> With the current code we make an assumption, that when the data _fork_ is present, we could read resource fork without checking for its presence. This simplification is losing that functionality, e.g. there is no way to check that an explicit data fork is present.
> 
> What do you think?

I think that the current code is wrong in its dichotomy of data fork being either present or not. Every mac file (except directory) has data fork. It may be zero-length but it's always there. The same as Unix or Windows plain files always have a stream associated with them, even if zero-length one. Our code is wrong in following aspects:
1) Ignoring zero-size data fork in stuffit.cpp
2) Not returning a zero-sized stream in openFileOrDataFork in a presence of just .rsrc file
3) Having hasDataFork at all. Successful return from openFileOrDataFork is all we need to confirm the file/data fork presence

File: base/commandLine.cpp (hunk 1 above)
Comment (phcoder) [reply base/commandLine.cpp:1448]: Yes. Plain file is just a datafork. This doesn't change. getDataFork has the same behaviour

File: engines/advancedDetector.cpp (hunk 2 above)
Comment (phcoder) [reply engines/advancedDetector.cpp:569]: Following check for non-null is sufficient. This doesn't change from getDataFork

**Comment by sev-** [comment]
> I think that the current code is wrong in its dichotomy of data fork being either present or not

That is a point which I was missing for all of these years. Thank you!

Regarding those three points you mention. Are you willing to fix those or shall I merge this first?

**Comment by phcoder** [comment]
> > I think that the current code is wrong in its dichotomy of data fork being either present or not
> 
> That is a point which I was missing for all of these years. Thank you!
> 
> Regarding those three points you mention. Are you willing to fix those or shall I merge this first?

1. Will be in a separate commit
2. Same
3. Fixed in this PR

**Comment by sev-** [comment]
Thank you!

**Pull request merged**